patch addrman bucketing, remove build

willcl-ark opened the pull request:
Closes: #212 

Yay, another PR which should be 3!

This turned out longer than I wanted, as after patching addrman bucketing I needed to update the images. These are currently based on a plethora of dockerfiles created using various scripts, with different versions for k8s and compose. By using the same image for both backends we can in the future drop our ipv4 generator code from compose and just have everything routable and bucketable on both. We also only need to rebuild half the number of images...

In undertaking this, I thought we could take the opportunity to switch to alpine-based images, as they're (much) smaller. This neccesitated some minor changes to `exec` methods in the backends as there is no `gosu` in alpine (but there is in LND pods, debian-based?)

Finally, the knock-on effect of this was breakage of the `build` test. In my opinion this should be removed altogether and we should require users to simply run:

```bash
warcli image build --registry=<my_registry>/<name> --repo=<gh_user>/<repo> --branch=<branch_name> --arches=amd64 --tag=<my_tag>
```

...and then supply image info to the graph. This is a bit more manual than auto-building, but it's not clear currently how that would work with k8s in any case. This ends up therefore removing compose-specific code only, and makes a few code sections much simpler. We could simplify even further by removing `version` and only permitting image, using our registry as the default. But I've left that for now.

In order to proceed with this change, if folks want things past the patch, we should first update https://hub.docker.com/r/bitcoindevproject/bitcoin-core with new unified (and patched) images, then remove the NO MERGE commit(s), both if we  want to keep image building on compose, otherwise only the penultimate.

We can then remove the following registries too for extra cleanliness:

https://hub.docker.com/r/bitcoindevproject/k8s-bitcoin-core
https://hub.docker.com/r/bitcoindevproject/warnet-bitcoin
https://hub.docker.com/r/bitcoindevproject/warnet-bitcoin-core

willcl-ark committed:
patch: addrman buckets ipv4 based on /32

willcl-ark commented:
Also worth considering: I cannot claim to fully understand the ramifications of this addrman patch. I've observed different buckets and positions being used rather than a single bucket, so I know that much works. However it's unclear to me yet whether there are any currently-unforseen consequences of bucketing by /32...

It occurred to me that we may have unwittingly altered the total number of buckets, but there may be other side-effects too.

josibake commented:
> Also worth considering: I cannot claim to fully understand the ramifications of this addrman patch. I've observed different buckets and positions being used rather than a single bucket, so I know that much works. However it's unclear to me yet whether there are any currently-unforseen consequences of bucketing by /32...

Agree, and I think we should clearly document this (at least for now). I still think this is the best route forward, but we should run a live warnet scenario for awhile and compare the raw addrman against some real nodes to get a sense of how closely the change mirrors mainnet. If this ends up introducing side effects, we'll need to go back to the drawing board, but for now this seems to unblock us.

willcl-ark commented:
FWIW, some image size comparisons:

Release image, current, debian-slim: 159MB
Built image, current, debian-slim: > 300MB!
(estimate built could be reduced to about 175MB but not tested)

Unified, proposed, alpine: 35.3MB

josibake commented on src/backends/compose/compose_backend.py:
```diff
@@ -135,8 +135,10 @@ def get_status(self, tank_index: int, service: ServiceType) -> RunningStatus:
                 return RunningStatus.PENDING
 
-    def exec_run(self, tank_index: int, service: ServiceType, cmd: str, user: str = "root") -> str:
+    def exec_run(self, tank_index: int, service: ServiceType, cmd: str, user="bitcoin") -> str:
```
in "images: unified alpine-based dockerfile" (https://github.com/bitcoin-dev-project/warnet/pull/261/commits/4dc71ff74d4267b9c3f45530c1490702bd81c188):

```suggestion
    def exec_run(self, tank_index: int, service: ServiceType, cmd: str, user: str = "bitcoin") -> str:
```

willcl-ark replied:
taken

josibake commented on src/backends/kubernetes/kubernetes_backend.py:
```diff
@@ -157,13 +157,12 @@ def get_status(self, tank_index: int, service: ServiceType) -> RunningStatus:
-        # because its installed in the bitcoin containers. we will need to rework
-        # this command if we decided to remove gosu from the containers
-        # TODO: change this if we remove gosu
+    def exec_run(self, tank_index: int, service: ServiceType, cmd: str, user="root"):
```
in "images: unified alpine-based dockerfile" (https://github.com/bitcoin-dev-project/warnet/commit/4dc71ff74d4267b9c3f45530c1490702bd81c188):

```suggestion
    def exec_run(self, tank_index: int, service: ServiceType, cmd: str, user: str = "root"):
```

willcl-ark replied:
taken

josibake commented on src/utils/image_build.py:
```diff
@@ -51,10 +47,9 @@ def build_and_upload_images(
 
     # Setup buildkit
-    builder_name = "warnet-builder"
+    builder_name = "bitcoind-builder"
```
in "fixup! images: unified alpine-based dockerfile" (https://github.com/bitcoin-dev-project/warnet/pull/261/commits/6070ea73f07dbccd0e08a1baec44ebe974f94e89):

in the previous commit this was called `warnet-bitcoind-builder`, not sure if that makes a difference. Also noticed you're removing the script in this commit, might be worth just dropping the script entirely from the PR?

willcl-ark replied:
Which script do you mean? This PR attempts to clean a load of them out already?

josibake replied:
I could have sworn you were adding `build-bitcoind.sh` in an earlier commit and then removing it in this one.. but I don't see it now? Either I'm losing my marbles or you already removed it haha

josibake commented:
> Finally, the knock-on effect of this was breakage of the build test. In my opinion this should be removed altogether and we should require users to simply run: <script>

I do agree this should be the normal mode of operation, but I don't think we should disable the build script altogether, in that I think it's still useful for CI. Needing to push images first to a registry before the CI can pass seems error prone and a PITA. AFAIK, if we are using minikube in the CI, we can build the images and use them directly in minikube, so maybe we keep the building and CI stuff to where it doesn't rely on hitting the registry? Or both: one that builds and tests the images, and another that runs with the prebuilt images, more as a regression test?

There may also be users who want to use warnet locally with docker or minikube that would benefit from the build script for local work.

willcl-ark commented:
> I do agree this should be the normal mode of operation, but I don't think we should disable the build script altogether, in that I think it's still useful for CI. Needing to push images first to a registry before the CI can pass seems error prone and a PITA.
> AFAIK, if we are using minikube in the CI, we can build the images and use them directly in minikube, so maybe we keep the building and CI stuff to where it doesn't rely on hitting the registry? Or both: one that builds and tests the images, and another that runs with the prebuilt images, more as a regression test?

Hmmm, I somewhat agree. let me see if the script can be modified to use `--load`, then we can use it to build and load into local registry, before using in minikube on CI. Then I'd argue that's equivalent to what we have now, and the rest can still go?


> There may also be users who want to use warnet locally with docker or minikube that would benefit from the build script for local work.

I think this would be addressed by the above too

josibake commented:
> Hmmm, I somewhat agree. let me see if the script can be modified to use --load, then we can use it to build and load into local registry, before using in minikube on CI. Then I'd argue that's equivalent to what we have now, and the rest can still go?

Sounds excellent!

willcl-ark commented:
Sorry for many rebases, somehow broke the scenarios test during the work...

willcl-ark committed:
ci: point minikube local registry to outer docker registry in CI

willcl-ark committed:
ci: disable k8s builder

This is not working currently, but can be re-enabled in the near future
(TM)

willcl-ark committed:
images: unified alpine-based dockerfile

willcl-ark committed:
remove k8s dockerfile

willcl-ark committed:
remove image-building scripts

willcl-ark committed:
remove default Dockerfile

willcl-ark committed:
rename alpine dockerfile

willcl-ark committed:
switch to root user exec for alpine

willcl-ark committed:
update image-builder for alpine

willcl-ark committed:
justfile: add build command using new Dockerfile

willcl-ark committed:
image: rehome image_build under cli

willcl-ark committed:
image: fixup image cli command

willcl-ark committed:
compose: add inline branch-build

willcl-ark committed:
k8s: add inline branch-build

But disable it for now in CI

willcl-ark committed:
server: lock during build stage to permit inline builds

willcl-ark committed:
cli: fix network start command return type

willcl-ark commented:
OK, split the big commits for easier review, and got all the tests passing again.

It sounds like we might now be waiting on @pinheadmz magic formula before considering merging this? Alternatively, we could merge this and simply update the (single) docker images, removing the patches and swapping with his (non-bitcoin) patches, if he gets it working.

> AFAIK, if we are using minikube in the CI, we can build the images and use them directly in minikube, so maybe we keep the building and CI stuff to where it doesn't rely on hitting the registry?

For now, I've disabled branch-building on k8s backend _in CI only_. It might work on your local machine though...

willcl-ark commented:
It also just occurred to me that with a unified image we could have a workflow to (re)build the docker images when we tag in this repo (or whatever)

willcl-ark commented on src/utils/image_build.py:
```diff
@@ -21,11 +21,7 @@ def build_and_upload_images(
-        build_args = (
-            '"--disable-tests --with-incompatible-bdb --without-gui --disable-bench '
-            "--disable-fuzz-binary --enable-suppress-external-warnings "
-            '--without-miniupnpc --without-natpmp"'
```
Don't need these as we don't install the deps for them in the image, so they don't get build automatically.

josibake commented on src/templates/entrypoint.sh:
```diff
@@ -16,20 +16,10 @@ if [ "${TOR:-0}" -eq 1 ]; then
     chown -R debian-tor:debian-tor /home/debian-tor
     # Start tor in the background
-    gosu debian-tor tor &
+    su-exec debian-tor:debian-tor tor &
```
do we still need su-exec, now that we are always running as the root user?

willcl-ark replied:
leaving for Tor (for now)

josibake commented on src/templates/Dockerfile_unified_alpine:
```diff
@@ -0,0 +1,86 @@
+    libzmq \
+    shadow \
+    sqlite-dev \
+    su-exec
```
can remove, yeah?

willcl-ark replied:
leaving for Tor (for now)

josibake commented on src/utils/image_build.py:
```diff
@@ -66,25 +61,27 @@ def build_and_upload_images(
         f" --tag {image_full_name}"
-        f" --file Dockerfile_k8 ."
+        f" --file src/templates/Dockerfile ."
         f" --push"
```
would be nicer if you added the {action} arg in this commit and then updated the justfile in the next, but not a big deal, was just very confused by the next commit haha

willcl-ark replied:
yeah. But rebasing is tRiCkY

josibake commented on src/cli/image.py:
```diff
@@ -17,11 +17,12 @@ def image():
-    Build bitcoind and bitcoin-cli from <repo>/<branch> and deploy to <registry> as <tag>
-    This requires docker and buildkit to be enabled.
+    Build bitcoind and bitcoin-cli from <repo>/<branch> as <registry>:<tag>.
+    Optionally deploy to remote registry using --action=push, otherwise image is loaded to local registry.
```
note for a follow up , would be nice if we gave examples in our docs of how to use this script, just to make it more clear

josibake commented on src/backends/kubernetes/kubernetes_backend.py:
```diff
@@ -18,6 +18,7 @@
 
```
tbh im not sure this will work on k8s. generally in k8s, the pattern for building images is you have a custom image with the correct permissions (Kaniko is the most popular one), so you pass of a build context to a container running in the cluster, and that container returns the image. this has some additional nice properties of allowing you to build multiple images in parallel (i think?). either way, i think this will take some more finessing to get working.

willcl-ark replied:
I think this will just have the RPC container build the image no?

But I can't see what exactly you're commenting on here (thank GH)

josibake reviewed (commented):
looks good, left a few comments

willcl-ark commented:
OK final force push done.

_Now_ if this gets approved, I'll update images before removing the top commit and merging

josibake commented:
ACK https://github.com/bitcoin-dev-project/warnet/commit/d3cc379f2b5e00c7bfa8abdedbaba0aea276cf1d

looks RFM, will let @m3dwards and @pinheadmz take a look first tho

willcl-ark committed:
images: use new registry with unified image

This image is based on alpine linux, and will work on k8s and compose
backends, as well as forming the base of custom builds.

One image to rule them all, and in the darkness bind them. Or something.

m3dwards commented:
ACK https://github.com/bitcoin-dev-project/warnet/commit/66444fff44256e45dc69feff8d28cf1cab9604e7

I think this is great.

josibake commented:
reACK https://github.com/bitcoin-dev-project/warnet/commit/66444fff44256e45dc69feff8d28cf1cab9604e7

diff looks good!

pinheadmz commented on .github/workflows/test.yml:
```diff
@@ -11,17 +11,17 @@ jobs:
+        backend: [compose]
     steps:
       - uses: actions/checkout@v3
-      - if: matrix.backend == 'compose'
```
Did you have to remove this "if"? If we can just keep everything the same and remove "k8s" from the matrix on tests we wanna skip, that'll be cleaner and easier to undo

willcl-ark replied:
When we add K8s back here, we'll likely need docker installed to build images. So we'd need to remove the if then. (Unless we find a weird way of building)

pinheadmz commented:
post post post post post post post post merge nit: https://github.com/bitcoin-dev-project/warnet/pull/376
